### PR TITLE
Modified api changes for in app logging

### PR DESCRIPTION
**Pull request by poojaRal**
**Taiga User Story:** None

**Taiga Task(s):** None

**What did you change?**
1. Added a new api -> survey_logger POST request to POST in-app logging timestamps for each completed survey.
2. Added a new table "survey_response_logger" to save the in-app logging activities
**How can we test?**
1. Test the api request 
http://localhost:3001/survey_logger
Sample Data for Testing :

``` json
{
	"loggerResults": [{
		"eventName": "check_surveys",
		"metaData": {
			"pin": "2001"
		},
		"startTime": 1458086022396
	}, {
		"eventName": "get_survey",
		"metaData": {
			"pin": "2001",
			"sid": 2
		},
		"startTime": 1458086023287
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "No Pain",
				"intensity": "0"
			}
		},
		"startTime": 1458086024437
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front abdominal",
				"intensity": "5"
			}
		},
		"startTime": 1458086026686
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front abdominal",
				"intensity": "5"
			}
		},
		"startTime": 1458086026696
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31
		},
		"startTime": 1458086028022
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1,
			"aid": 4
		},
		"startTime": 1458086028492
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1
		},
		"startTime": 1458086028917
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2,
			"aid": 9
		},
		"startTime": 1458086029277
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2
		},
		"startTime": 1458086029657
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3,
			"aid": 13
		},
		"startTime": 1458086030031
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3
		},
		"startTime": 1458086030723
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4,
			"aid": 20
		},
		"startTime": 1458086031105
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4
		},
		"startTime": 1458086031719
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5,
			"aid": 23
		},
		"startTime": 1458086032234
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5
		},
		"startTime": 1458086032713
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5
		},
		"startTime": 1458086033243
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5,
			"aid": 23
		},
		"startTime": 1458086033258
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4
		},
		"startTime": 1458086033415
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4,
			"aid": 20
		},
		"startTime": 1458086033431
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3
		},
		"startTime": 1458086033568
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3,
			"aid": 13
		},
		"startTime": 1458086033583
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2
		},
		"startTime": 1458086033725
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2,
			"aid": 9
		},
		"startTime": 1458086033737
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1
		},
		"startTime": 1458086033919
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1,
			"aid": 4
		},
		"startTime": 1458086033933
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31
		},
		"startTime": 1458086034082
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front abdominal",
				"intensity": "5"
			}
		},
		"startTime": 1458086034083
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front abdominal",
				"intensity": "5"
			}
		},
		"startTime": 1458086034084
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front abdominal",
				"intensity": "5"
			}
		},
		"startTime": 1458086034086
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front left arm",
				"intensity": "5"
			}
		},
		"startTime": 1458086034262
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31,
			"aid": {
				"location": "front left arm",
				"intensity": "5"
			}
		},
		"startTime": 1458086034274
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 31
		},
		"startTime": 1458086036013
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1,
			"aid": 4
		},
		"startTime": 1458086036031
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 1
		},
		"startTime": 1458086036968
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2,
			"aid": 9
		},
		"startTime": 1458086036984
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 2
		},
		"startTime": 1458086037374
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3,
			"aid": 13
		},
		"startTime": 1458086037391
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 3
		},
		"startTime": 1458086037585
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4,
			"aid": 20
		},
		"startTime": 1458086037600
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 4
		},
		"startTime": 1458086037756
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5,
			"aid": 23
		},
		"startTime": 1458086037774
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5
		},
		"startTime": 1458086037948
	}, {
		"eventName": "previous_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5
		},
		"startTime": 1458086039178
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5,
			"aid": 23
		},
		"startTime": 1458086039194
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 5
		},
		"startTime": 1458086039731
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 6,
			"aid": 30
		},
		"startTime": 1458086040116
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 6
		},
		"startTime": 1458086040586
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 8,
			"aid": 114
		},
		"startTime": 1458086040919
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 8
		},
		"startTime": 1458086041690
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 9,
			"aid": 119
		},
		"startTime": 1458086042029
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 9
		},
		"startTime": 1458086042640
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 10,
			"aid": 125
		},
		"startTime": 1458086042955
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 10
		},
		"startTime": 1458086043443
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 12,
			"aid": 133
		},
		"startTime": 1458086043797
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 12
		},
		"startTime": 1458086044555
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 14,
			"aid": 145
		},
		"startTime": 1458086044931
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 14
		},
		"startTime": 1458086045348
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 15,
			"aid": 149
		},
		"startTime": 1458086045693
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 15
		},
		"startTime": 1458086046078
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 16,
			"aid": 38
		},
		"startTime": 1458086046502
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 16
		},
		"startTime": 1458086047184
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 19,
			"aid": 55
		},
		"startTime": 1458086047521
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 19
		},
		"startTime": 1458086048016
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 20,
			"aid": 60
		},
		"startTime": 1458086048378
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 20
		},
		"startTime": 1458086048957
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 21,
			"aid": 64
		},
		"startTime": 1458086049317
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 21
		},
		"startTime": 1458086049828
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 22,
			"aid": 70
		},
		"startTime": 1458086050207
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 22
		},
		"startTime": 1458086050784
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 23,
			"aid": 75
		},
		"startTime": 1458086051177
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 23
		},
		"startTime": 1458086051556
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 25,
			"aid": 85
		},
		"startTime": 1458086052189
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 25
		},
		"startTime": 1458086052792
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 26,
			"aid": 89
		},
		"startTime": 1458086053573
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 26
		},
		"startTime": 1458086054052
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 27,
			"aid": 95
		},
		"startTime": 1458086054585
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 27
		},
		"startTime": 1458086055040
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 28,
			"aid": 99
		},
		"startTime": 1458086055425
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 28
		},
		"startTime": 1458086056167
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 29,
			"aid": 105
		},
		"startTime": 1458086056886
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 29
		},
		"startTime": 1458086057234
	}, {
		"eventName": "answer_option_selected",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 30,
			"aid": 108
		},
		"startTime": 1458086057630
	}, {
		"eventName": "next_btn_click",
		"metaData": {
			"pin": "2001",
			"sid": 2,
			"qid": 30
		},
		"startTime": 1458086058176
	}, {
		"eventName": "submit_survey",
		"metaData": {
			"pin": "2001",
			"sid": "2"
		},
		"startTime": 1458086071006
	}]
}
```
Check whether all the entries from this JSON are saved into "survey_logger" table in the database.

### CODE REVIEW

File: api/router.js [hunk 1]
```diff
@@ -86,5 +87,43 @@ module.exports = [
...
+                                    .number()
+                                    .integer()
+                                    .positive(),
+                                qid: Joi
+                                    .number()
+                                    .integer()
+                                    .positive(),
+                                aid: Joi
+                                    .alternatives()
+                                    .try(Joi.number(), Joi.object())
```
Comment (ChristianMurphy) [inline api/router.js]: could we focus the validation of `aid` a bit more

something like
``` javascript
aid: Joi
     .alternatives()
    .try(
        Joi
            .number()
            .integer()
            .positive(),
        Joi
            .object({
                location: Joi.string(),
                intensity: Joi.number().min(minBodyPainIntensity).max(maxBodyPainIntensity)
            })
    )
```

File: model/survey-logger.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    sequelize.define(
+        'survey_logger',
+        {
+            eventName: {
+                type: Sequelize.STRING
+            },
+            metaData: {
+                type: Sequelize.STRING
+            },
+            startTime: {
```
Comment (ChristianMurphy) [inline model/survey-logger.js]: nit `startTime` -> `start` :art:

File: api/helper/process-logger-results.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Takes in a survey_instance model and processes it into a format recognized by app.
+ * @param {Object} logs - list of objects of survey result logs
+ * @returns {Object} processed list of logs in an array
+ */
+function processLoggerResults (logs) {
+    return {
+        eventName: logs.eventName,
+        metaData: JSON.stringify(logs.metaData),
```
Comment (ChristianMurphy) [inline api/helper/process-logger-results.js]: Question: what (if any) limit will we have on JSON size?

File: model/survey-logger.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * @typedef {String} eventName - The event captured from the app
+ * @typedef {String} metaData - The metadata - PIN, SurveyInstanceID, Question ID, Answer ID related to the event
+ * @typedef {Date} startTime - The timestamp captured when this event occurred in the app.
+ * @returns {Null} nothing
+ */
+function register (sequelize) {
+    /**
+     * Each QuestionResult will be associated with a SurveyInstance and a QuestionOption
+     * This represents a single response to a Survey QuestionTemplate
+     * @typedef {Object} SurveyLogger
```
Comment (ChristianMurphy) [inline model/survey-logger.js]: The `@typedef`s lines 12-14 should be moved under this `@typedef` and renamed as `@property`s

File: api/helper/process-logger-results.js (hunk 3 above)
Comment (poojaRal) [inline api/helper/process-logger-results.js]: As of now we cannot determine what the size limit will be. The app will log all button interactions any number of times a user clicks. It is difficult to judge what would be the interaction pattern at this point.

File: api/router.js (hunk 1 above)
Comment (poojaRal) [inline api/router.js]: Sure. Will make changes in the code accordingly.

File: model/survey-logger.js (hunk 4 above)
Comment (poojaRal) [inline model/survey-logger.js]: Yup. Will make changes.

File: model/survey-logger.js (hunk 2 above)
Comment (poojaRal) [inline model/survey-logger.js]: OK.

File: api/helper/process-logger-results.js (hunk 3 above)
Comment (ChristianMurphy) [inline api/helper/process-logger-results.js]: Right, but the metaData field in the database is a string, which means it is either a `CHAR` or a `VARCHAR` both have limits.

**Comment by ChristianMurphy** [comment]
LGTM

**Comment by iamaureen** [comment]
:shipit:

**Pull request merged**